Add support for utf16 high/low surrogates in measurestring/drawstring.

bmerlet opened the pull request:
This patch allows Garphics.MeasureString() and Graphics.DrawString() to correctly measure and draw unicode characters with a codepoint greater than 0xffff.

filipnavara commented:
The change looks ok, but the name of the function (ucs2_to_utf8) is a bit misleading now since the difference between `ucs2` and `utf16` is the introduction of surrogate pairs.

bmerlet commented:
Is this a request for me to change the name of the function?
I don't mind doing it, I am just not sure how to interpret the comment.

filipnavara commented:
Sorry, I should have been clearer :) I'd probably change the name to make it more obvious what the function does.

dnfclas commented:
[![CLA assistant check](https://cla.dotnetfoundation.org/pull/badge/not_signed)](https://cla.dotnetfoundation.org/mono/libgdiplus?pullRequest=549) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/mono/libgdiplus?pullRequest=549) before we can accept your contribution.<br/><br/>:x: bmerlet [sign now](https://cla.dotnetfoundation.org/mono/libgdiplus?pullRequest=549)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/mono/libgdiplus?pullRequest=549) it.</sub>

bmerlet commented:
So I renamed the function (to utf16_to_utf8()). I also fixed utf8_length_for_uscs2_string() to add the correct number of bytes for utf16 surrogates, and renamed it as well. Finally, I looked through the code for other uses of the 0xe000 and 0xd800 constants, found another one. They are used by utf8_to_ucs2(). This function is only used to translate font names from utf8 back to ucs2. I decided to leave utf8_to_ucs2() alone. Let me know if you think it ought to be upgraded as well.

bmerlet commented:
When are the Jenkins build supposed to run? It's been 5 days since this patch got approved...

filipnavara commented:
You need to be on a whitelist for Jenkins to run. /cc @akoeplinger

directhex commented:
@monojenkins build Mono Jenkins Linux

directhex commented:
@monojenkins build Mono Jenkins Windows
